File: benchmark/synduce/tailopt/mps.f
Config ExtraGrammar = "AutoLifter";
Inductive List = nil Unit | cons {Int, List};

max = \x: Int. \y: Int. if > x y then x else y;

spec = \xs: List. (fix (
  \f: List -> {Int, Int}. \xs: List.
  match xs with
    nil _ -> {0,0}
  | cons {h, t} ->
    let r = f t in
    {+ h r.1, max 0 (+ h r.2)}
  end
) xs).2;

snoc = fix (
  \f: List -> Int -> List. \xs: List. \w: Int.
  match xs with
    nil _ -> cons {w, nil unit}
  | cons {h, t} -> cons {h, f t w}
  end
);

repr = (fix (
  \f: Compress List -> List -> Compress List.
  \pre: Compress List. \xs: List.
  match xs with
    nil _ -> pre
  | cons {h, t} -> f (snoc pre h) t
  end
)) (nil unit);

main = \xs: List. spec (repr xs);

File: ml-benchmark/synduce/tailopt/mps.ml
type t = Nil of unit | Cons of int * t

let max x y = if x > y then x else y

let spec xs =
  let rec f xs =
    match xs with
    | Nil _ -> (0, 0)
    | Cons (h, t) ->
      let r = f t in
      match r with
      | (a, b) -> (h + a, max 0 (h + b))
  in
  match f xs with
  | (_, b) -> b

let rec snoc xs w =
  match xs with
  | Nil _ -> Cons (w, Nil ())
  | Cons (h, t) -> Cons (h, snoc t w)

let repr =
  let rec f pre xs =
    match xs with
    | Nil _ -> pre
    | Cons (h, t) -> f (snoc pre h) t
  in
  fun xs -> f (Nil ()) xs

let program xs = spec (repr xs)